### PR TITLE
Oceanwater 1078

### PR DESCRIPTION
**Pull request by Keeganfn**
## Summary
Testing and fixes for fault handling pattern plans 1-6.

- Fixed Panning code in plans 1-6 to not error out with new antenna limits.
- Fixed logic in plans 1 and 2 to allow for repeating actions after a fault occurs during or before execution.
- Changed PanTiltMoveJoints() -> Pan() in plans 1-6 for simplicity and better fault checking.
- Changed Pan.plp and Tilt.plp to use their specific fault lookup instead of the general AntennaFault lookup.
- Updated documentation on all plans to better describe the current behaviors. Particularly plans 4-6. 

## Notes on Code
All plans now run to completion, however only plans 1 and 2 work completely as intended. Plans 3-6 were created when PLEXIL did not have access to as much fault introspection, and some workarounds were needed. They do not always demonstrate the desired behavior which I noted in the comments above each file. 

The plans themselves are very useful examples. Once more fault work has been done (health monitor, goal faults, etc) I would like to revisit and update them.

I think there are two options for merging the FaultPatterns into  release 11-1: 
1. Merge all fault handling patterns, with my additional warnings/notes to serve as example plans. 
2. Only merge plans 1 and 2, and merge plans 3-6 at a later date once they are updated with the new proposed fault tools. 

I will also create a JIRA ticket to update plans 3-6. 

## Build
1. ow_simulator should be on master branch, as this change will be merged into release 11-1
3. ```catkin clean```
4. ```catkin build```

## Test
For all plans: 
1. ``` roslaunch ow europa_terminator_workspace.launch ```
2. ``` roslaunch ow_plexil ow_exec.launch plan:="FaultHandlingPattern1.plx" ``` (Or you can use the GUI, replacing 1 with 
whatever plan you would like to run)
3. Use the Dynamic reconfigure GUI to insert ant_pan_joint_locked_failure, and any arm faults. For FaultPattern6, you can trigger a low batter fault by setting power draw to its highest value, and triggering the backup action until it triggers.

### FaultPattern1: 
**Original Desired Behavior**: Pan in 15 degree increments as long as no antenna pan fault is present.  If an antenna pan fault occurs, pause until the fault is removed.  This plan must be terminated with an interrupt (e.g. Control-C).
**Actual Behavior:** Works correctly. 

### FaultPattern2: 
**Original Desired Behavior**: Panning and arm unstowing/stowing as long as no relevant fault is present. An arm fault will not prevent the arm from moving and vice versa. 
**Actual Behavior:** Works correctly. 

### FaultPattern3: 
**Original Desired Behavior:** Pan to 180 degrees, unstow arm, and then stow. If a relevant fault is present the action will be paused and then skipped once the fault is resolved. 

**Actual Behavior:** Pan will always complete its action (this is due to how the action is set up). If an arm fault occurs during stow or unstow the plan will end. 

### FaultPattern4: 
**Original Desired Behavior:** Pan to 180 degrees, unstow arm, and then stow. If a relevant fault is detected, pause that action and then resume once the fault is cleared. 

**Actual Behavior:** Unstow will not always resume after a fault is cleared. If arm fault occurs during stow the plan will end. 

### FaultPattern5: 
**Original Desired Behavior:** Pan to 180 degrees, unstow arm, and then stow. If a relevant fault is detected, pause that action and start a backup pan action (unless the fault is an antenna fault) until the fault is cleared, where it will then resume the original action.

**Actual Behavior:** Unstow will not always resume after a fault is cleared. If arm fault occurs during stow the plan will end. 

### FaultPattern6: 
**Original Desired Behavior:** Pan to 180 degrees, unstow arm, and then stow. If a relevant fault is detected, pause that action and start a backup pan action (unless the fault is an antenna fault) until the fault is cleared, where it will then resume the original action. This also has additional checks for battery life faults which work as intended.

**Actual Behavior:** Unstow will not always resume after a fault is cleared. If arm fault occurs during stow the plan will end.

Thanks!

### CODE REVIEW

File: ow_plexil/src/plans/FaultHandlingPattern2.plp [hunk 1]
```diff
@@ -25,15 +25,26 @@ FaultHandlingPattern2:
...
-      LibraryCall PanTiltMoveJoints (PanDegrees=NewAngle, TiltDegrees=0);
+      if NewAngle > 180{
+        NewAngle = NewAngle - 360;
+      }
+
+      LibraryCall Pan (Degrees=NewAngle);
     }
+  }
 
     Pattern2UnstowStow:
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern2.plp:49]: Small improvement: the sequence of Unstow and Stow in this node can be replaced by the new `LibraryCall StowSequence()`.

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 2]
```diff
@@ -18,14 +18,25 @@ FaultHandlingPattern1:
 
   LibraryCall InitializeAntennaAndArm ();
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:19]: This initialization takes a while, is not always all needed, and it's hard to tell when the "pattern" starts.  If one injects faults during the initialization, behavior is unspecified.

I suggest doing away with the `InitializeAntennaAndArm` plan, and just have each fault pattern plan perform the initialization it needs, if any.  The initializing segment should be enclosed in messages, so that one knows when the actual pattern has started.

File: ow_plexil/src/plans/FaultHandlingPattern3.plp [hunk 3]
```diff
@@ -5,11 +5,11 @@
 // If no faults are occurring, this plan will pan 180 degrees,
 // and then unstow and stow the arm.
 
-// If any fault is injected, the plan will pause and wait for resolution.
-// Resolution is defined as the removal of the fault.
+// Only the AntennaPan action will hault for resolution,
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern3.plp:8]: Typo `hault` but see my overall comment on plans 3-6, coming shortly.

**Comment by kmdalal** [comment]
Patterns 1 and 2 are good to go.  Patterns 3-6 are confusing and not fully working, as mentioned in the PR description.  In particular, their file comments are hard to follow.  Still, the improvements made to them are good, and the code could be useful as examples.

I suggest including 3-6 in this merge, but clean up their comments as much as possible, and put a warning at the top of each saying that this plan cannot work as intended until some fault/action support is added in the simulator.  

Tackling them as a new Jira makes sense, and the patterns could then also be improved or modified.

**Comment by Keeganfn** [comment]
Per review comments:
- Plan2 improvement to use StowSequence. 
- Got rid of unnecessary initialization to speed up plans.
- Fixed "hault" typo. 
- Added warnings and made documentation more clear at the top of each file.

**Review by kmdalal — APPROVED** [review]
Looks good.  Nice work!
